File: modules/algorithmanalysis/AlgorithmsModule01.java
/*
 * CSC310 — Module 01: Algorithm Analysis
 * Sample implementations in Java.
 *
 * Algorithms covered:
 *   - Linear Search          O(n) worst
 *   - Binary Search          O(log n) worst (iterative & recursive)
 *   - Insertion Sort         O(n^2) worst, O(n) best   (CLRS Fig. 2.1)
 *   - Merge Sort             Θ(n log n)                (CLRS §2.3)
 *   - Towers of Hanoi        Θ(2^n)
 *   - Recursive vs iterative factorial (Θ(n) stack vs Θ(1) stack)
 *   - Empirical timing demo connecting asymptotic to wall-clock
 *
 * The race-condition example is in RaceConditionDemo.java so it can be
 * compiled and run separately.
 *
 * Compile: javac AlgorithmsModule01.java
 * Run    : java  AlgorithmsModule01
 */

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class AlgorithmsModule01 {

    // ---------------------------------------------------------------------
    // 1. LINEAR SEARCH
    // Best Θ(1), worst/avg Θ(n), space Θ(1)
    // ---------------------------------------------------------------------
    public static SearchResult linearSearch(int[] arr, int target) {
        long comparisons = 0;
        for (int i = 0; i < arr.length; i++) {
            comparisons++;                  // key processing step
            if (arr[i] == target) return new SearchResult(i, comparisons);
        }
        return new SearchResult(-1, comparisons);
    }

    // ---------------------------------------------------------------------
    // 2. BINARY SEARCH — requires sorted input (a critical assumption!)
    // Best Θ(1), worst Θ(log n). Iterative is Θ(1) space; recursive Θ(log n).
    // ---------------------------------------------------------------------
    public static SearchResult binarySearchIterative(int[] arr, int target) {
        long comparisons = 0;
        int low = 0, high = arr.length - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;   // avoids overflow vs (low+high)/2
            comparisons++;
            if (arr[mid] == target) return new SearchResult(mid, comparisons);
            if (arr[mid] < target) low = mid + 1;
            else                   high = mid - 1;
        }
        return new SearchResult(-1, comparisons);
    }

    public static SearchResult binarySearchRecursive(int[] arr, int target) {
        return binarySearchRecursiveHelper(arr, target, 0, arr.length - 1, 0);
    }

    private static SearchResult binarySearchRecursiveHelper(
            int[] arr, int target, int low, int high, long cmp) {
        if (low > high) return new SearchResult(-1, cmp);
        int mid = low + (high - low) / 2;
        cmp++;
        if (arr[mid] == target) return new SearchResult(mid, cmp);
        if (arr[mid] < target)
            return binarySearchRecursiveHelper(arr, target, mid + 1, high, cmp);
        return binarySearchRecursiveHelper(arr, target, low, mid - 1, cmp);
    }

    // ---------------------------------------------------------------------
    // 3. INSERTION SORT (CLRS Figure 2.1, translated to 0-based indexing)
    // Best  Θ(n)   — already sorted
    // Worst Θ(n^2) — reverse sorted
    // Space Θ(1)   — in-place
    // ---------------------------------------------------------------------
    public static long insertionSort(int[] A) {
        long comparisons = 0;
        // CLRS: for j = 2 to A.length   (1-based)
        // Java: for j = 1 to A.length-1 (0-based)
        for (int j = 1; j < A.length; j++) {
            int key = A[j];
            int i = j - 1;
            while (i >= 0) {
                comparisons++;
                if (A[i] <= key) break;
                A[i + 1] = A[i];
                i--;
            }
            A[i + 1] = key;
        }
        return comparisons;
    }

    // ---------------------------------------------------------------------
    // 4. MERGE SORT — divide and conquer
    // T(n) = 2 T(n/2) + Θ(n)  →  Θ(n log n)
    // Auxiliary space Θ(n)
    // ---------------------------------------------------------------------
    public static long mergeSort(int[] A) {
        long[] comparisons = {0};
        int[] aux = new int[A.length];
        mergeSortHelper(A, aux, 0, A.length - 1, comparisons);
        return comparisons[0];
    }

    private static void mergeSortHelper(int[] A, int[] aux, int lo, int hi,
                                        long[] comparisons) {
        if (lo >= hi) return;
        int mid = lo + (hi - lo) / 2;
        mergeSortHelper(A, aux, lo, mid, comparisons);
        mergeSortHelper(A, aux, mid + 1, hi, comparisons);
        merge(A, aux, lo, mid, hi, comparisons);
    }

    private static void merge(int[] A, int[] aux, int lo, int mid, int hi,
                              long[] comparisons) {
        // Copy A[lo..hi] into aux[lo..hi]
        for (int k = lo; k <= hi; k++) aux[k] = A[k];

        int i = lo, j = mid + 1;
        for (int k = lo; k <= hi; k++) {
            if      (i > mid)              A[k] = aux[j++];
            else if (j > hi)               A[k] = aux[i++];
            else {
                comparisons[0]++;
                if (aux[j] < aux[i]) A[k] = aux[j++];
                else                 A[k] = aux[i++];
            }
        }
    }

    // ---------------------------------------------------------------------
    // 5. TOWERS OF HANOI
    // T(n) = 2 T(n-1) + 1  →  T(n) = 2^n - 1     i.e.  Θ(2^n)
    // Call-stack depth Θ(n)
    // ---------------------------------------------------------------------
    public static List<int[]> towersOfHanoi(int n) {
        List<int[]> moves = new ArrayList<>();
        hanoi(n, 'A', 'C', 'B', moves);
        return moves;
    }

    private static void hanoi(int n, char from, char to, char aux,
                              List<int[]> moves) {
        if (n == 1) { moves.add(new int[]{from, to}); return; }
        hanoi(n - 1, from, aux, to, moves);
        moves.add(new int[]{from, to});
        hanoi(n - 1, aux, to, from, moves);
    }

    // ---------------------------------------------------------------------
    // 6. RECURSIVE vs ITERATIVE — same Θ(n) time, different stack space
    // ---------------------------------------------------------------------
    public static long factorialRecursive(int n) {
        if (n <= 1) return 1L;
        return n * factorialRecursive(n - 1);     // Θ(n) stack frames
    }

    public static long factorialIterative(int n) {
        long result = 1;
        for (int k = 2; k <= n; k++) result *= k; // Θ(1) extra space
        return result;
    }

    // ---------------------------------------------------------------------
    // 7. EMPIRICAL GROWTH DEMO
    // ---------------------------------------------------------------------
    private static void growthDemo() {
        // JIT warm-up: the HotSpot compiler optimizes hot methods after
        // ~10K invocations. Without warm-up the first row is misleadingly
        // slow because we're measuring interpreter speed.
        int[] warm = new int[500];
        for (int rep = 0; rep < 5; rep++) {
            for (int i = 0; i < warm.length; i++) warm[i] = warm.length - i;
            insertionSort(warm);
            for (int i = 0; i < warm.length; i++) warm[i] = warm.length - i;
            mergeSort(warm);
        }

        System.out.printf("%8s | %14s | %10s | %9s%n",
                "n", "insertion (s)", "merge (s)", "ins/merge");
        System.out.println("-".repeat(55));
        for (int n : new int[]{1_000, 2_000, 4_000, 8_000}) {
            int[] worstIns = new int[n];
            for (int i = 0; i < n; i++) worstIns[i] = n - i;       // reversed
            int[] worstMrg = worstIns.clone();

            long t0 = System.nanoTime();
            insertionSort(worstIns);
            double tIns = (System.nanoTime() - t0) / 1e9;

            t0 = System.nanoTime();
            mergeSort(worstMrg);
            double tMrg = (System.nanoTime() - t0) / 1e9;

            System.out.printf("%8d | %14.4f | %10.4f | %8.1fx%n",
                    n, tIns, tMrg, tIns / Math.max(tMrg, 1e-9));
        }
    }

    // ---------------------------------------------------------------------
    // Demo harness
    // ---------------------------------------------------------------------
    public static void main(String[] args) {
        System.out.println("=== Linear vs Binary Search ===");
        int n = 1_000_000;
        int[] sortedData = new int[n];
        for (int i = 0; i < n; i++) sortedData[i] = i + 1;
        int target = 987_654;

        SearchResult lin = linearSearch(sortedData, target);
        SearchResult bin = binarySearchIterative(sortedData, target);
        System.out.printf("Linear search comparisons : %,d%n", lin.comparisons);
        System.out.printf("Binary search comparisons : %,d  (log2 n ≈ %.1f)%n",
                bin.comparisons, Math.log(n) / Math.log(2));

        System.out.println("\n=== Insertion Sort: best vs worst ===");
        int[] sortedIn = new int[1000];
        int[] reversedIn = new int[1000];
        for (int i = 0; i < 1000; i++) { sortedIn[i] = i + 1; reversedIn[i] = 1000 - i; }
        System.out.printf("Sorted  (best)   comparisons: %,7d   ≈ n - 1 = 999%n",
                insertionSort(sortedIn));
        System.out.printf("Reverse (worst)  comparisons: %,7d   ≈ n(n-1)/2 = %,d%n",
                insertionSort(reversedIn), 1000 * 999 / 2);

        System.out.println("\n=== Merge Sort ===");
        List<Integer> list = new ArrayList<>();
        for (int i = 0; i < 1000; i++) list.add(i);
        Collections.shuffle(list, new Random(42));
        int[] arr = list.stream().mapToInt(Integer::intValue).toArray();
        long cmp = mergeSort(arr);
        System.out.printf("Merge sort comparisons on n=1000: %,d  (n log2 n ≈ %,d)%n",
                cmp, (int) (1000 * Math.log(1000) / Math.log(2)));
        assert isSorted(arr);

        System.out.println("\n=== Towers of Hanoi ===");
        for (int k : new int[]{1, 3, 5, 10}) {
            int moves = towersOfHanoi(k).size();
            System.out.printf("n = %2d : %,5d moves   (2^n - 1 = %,d)%n",
                    k, moves, (1 << k) - 1);
        }

        System.out.println("\n=== Recursive vs Iterative Factorial ===");
        System.out.printf("factorialIterative(10) = %d%n", factorialIterative(10));
        System.out.printf("factorialRecursive(10) = %d%n", factorialRecursive(10));
        System.out.println("Recursive version uses Θ(n) stack frames.");

        System.out.println("\n=== Empirical Growth Demo ===");
        growthDemo();
    }

    private static boolean isSorted(int[] a) {
        for (int i = 1; i < a.length; i++) if (a[i - 1] > a[i]) return false;
        return true;
    }

    // ---------------------------------------------------------------------
    static class SearchResult {
        final int index;
        final long comparisons;
        SearchResult(int index, long comparisons) {
            this.index = index;
            this.comparisons = comparisons;
        }
        @Override public String toString() {
            return "SearchResult{index=" + index + ", comparisons=" + comparisons + "}";
        }
    }
}

File: modules/algorithmanalysis/algorithms_module01.py
"""
CSC310 — Module 01: Algorithm Analysis
Sample implementations in Python.

Covers every algorithm referenced in the module notes:
    - Linear Search                       O(n)        worst case
    - Binary Search (iterative+recursive) O(log n)    worst case
    - Insertion Sort                      O(n^2)      worst case, O(n) best
    - Merge Sort                          Θ(n log n)  all cases
    - Towers of Hanoi                     Θ(2^n)
    - Recursive vs iterative factorial    illustrates stack-space difference
    - Empirical timing demo               connects asymptotic to wall-clock
    - Race-condition demonstration        ties Module 01 to real-world threads

Each function counts its key processing steps (the "basic operation" from
the RAM model in the notes) so we can observe the growth function in action.

Note on indexing: CLRS pseudocode is 1-indexed; Python is 0-indexed.
The algorithms are translated to 0-based indexing here.
"""

from __future__ import annotations
import math
import random
import threading
import time
from typing import Callable


# ---------------------------------------------------------------------------
# 1. LINEAR SEARCH
# ---------------------------------------------------------------------------
# Best:    Θ(1)   — target at index 0
# Average: Θ(n)   — target equally likely anywhere
# Worst:   Θ(n)   — target absent or at last index
# Space:   Θ(1)   — only a loop variable
def linear_search(arr: list, target) -> tuple[int, int]:
    """Return (index_or_-1, comparisons_performed)."""
    comparisons = 0
    for i, value in enumerate(arr):
        comparisons += 1                     # the key processing step
        if value == target:
            return i, comparisons
    return -1, comparisons


# ---------------------------------------------------------------------------
# 2. BINARY SEARCH (requires sorted input — an assumption that must hold!)
# ---------------------------------------------------------------------------
# Best:  Θ(1)      — target is the first midpoint
# Worst: Θ(log n)  — search space halved each iteration
# Space: Θ(1) iterative / Θ(log n) recursive (stack frames)
def binary_search_iterative(arr: list, target) -> tuple[int, int]:
    comparisons = 0
    low, high = 0, len(arr) - 1
    while low <= high:
        mid = (low + high) // 2
        comparisons += 1
        if arr[mid] == target:
            return mid, comparisons
        if arr[mid] < target:
            low = mid + 1
        else:
            high = mid - 1
    return -1, comparisons


def binary_search_recursive(arr: list, target, low: int = 0,
                            high: int | None = None,
                            comparisons: int = 0) -> tuple[int, int]:
    if high is None:
        high = len(arr) - 1
    if low > high:
        return -1, comparisons
    mid = (low + high) // 2
    comparisons += 1
    if arr[mid] == target:
        return mid, comparisons
    if arr[mid] < target:
        return binary_search_recursive(arr, target, mid + 1, high, comparisons)
    return binary_search_recursive(arr, target, low, mid - 1, comparisons)


# ---------------------------------------------------------------------------
# 3. INSERTION SORT  (CLRS Figure 2.1)
# ---------------------------------------------------------------------------
# Best:  Θ(n)   — already sorted; inner while runs 0 times
# Worst: Θ(n^2) — reverse sorted; inner while runs j-1 times
# Space: Θ(1)  — sorts in-place
def insertion_sort(arr: list) -> int:
    """Sort in-place. Return number of comparisons."""
    comparisons = 0
    # CLRS uses j = 2..length (1-based); Python equivalent is j = 1..len(arr)-1
    for j in range(1, len(arr)):
        key = arr[j]
        i = j - 1
        while i >= 0:
            comparisons += 1
            if arr[i] <= key:
                break
            arr[i + 1] = arr[i]
            i -= 1
        arr[i + 1] = key
    return comparisons


# ---------------------------------------------------------------------------
# 4. MERGE SORT  (divide-and-conquer; CLRS §2.3)
# ---------------------------------------------------------------------------
# Time:  Θ(n log n) in all cases — recurrence T(n) = 2T(n/2) + Θ(n)
# Space: Θ(n)      — temporary arrays for merging (auxiliary space)
def merge_sort(arr: list) -> tuple[list, int]:
    """Return (sorted_copy, comparisons)."""
    comparisons = [0]

    def _merge(left: list, right: list) -> list:
        merged, i, j = [], 0, 0
        while i < len(left) and j < len(right):
            comparisons[0] += 1
            if left[i] <= right[j]:
                merged.append(left[i]); i += 1
            else:
                merged.append(right[j]); j += 1
        merged.extend(left[i:])
        merged.extend(right[j:])
        return merged

    def _sort(a: list) -> list:
        if len(a) <= 1:
            return a
        mid = len(a) // 2
        return _merge(_sort(a[:mid]), _sort(a[mid:]))

    return _sort(arr), comparisons[0]


# ---------------------------------------------------------------------------
# 5. TOWERS OF HANOI
# ---------------------------------------------------------------------------
# Recurrence: T(n) = 2 T(n - 1) + 1  with T(1) = 1   →   T(n) = 2^n - 1
# Time:  Θ(2^n)      — exponential blowup illustrates why doubling CPU speed
#                      lets us solve only one more disk
# Space: Θ(n) stack — call stack depth equals the number of disks
def towers_of_hanoi(n: int, source: str = "A", target: str = "C",
                    auxiliary: str = "B",
                    moves: list | None = None) -> list:
    if moves is None:
        moves = []
    if n == 1:
        moves.append((source, target))
        return moves
    towers_of_hanoi(n - 1, source, auxiliary, target, moves)
    moves.append((source, target))
    towers_of_hanoi(n - 1, auxiliary, target, source, moves)
    return moves


# ---------------------------------------------------------------------------
# 6. RECURSIVE vs ITERATIVE — illustrates Θ(n) stack vs Θ(1) stack
# ---------------------------------------------------------------------------
def factorial_recursive(n: int) -> int:
    """Θ(n) time, Θ(n) stack space. Risk of RecursionError for large n."""
    if n <= 1:
        return 1
    return n * factorial_recursive(n - 1)


def factorial_iterative(n: int) -> int:
    """Θ(n) time, Θ(1) extra space."""
    result = 1
    for k in range(2, n + 1):
        result *= k
    return result


# ---------------------------------------------------------------------------
# 7. EMPIRICAL TIMING DEMO
# ---------------------------------------------------------------------------
def time_algorithm(fn: Callable, *args) -> float:
    start = time.perf_counter()
    fn(*args)
    return time.perf_counter() - start


def growth_demo() -> None:
    """Show that wall-clock time tracks the predicted growth function."""
    print(f"{'n':>8} | {'insertion (s)':>14} | {'merge (s)':>10} | "
          f"{'ins/merge':>9}")
    print("-" * 55)
    for n in (1_000, 2_000, 4_000, 8_000):
        # Worst case for insertion sort: reverse-sorted input
        worst = list(range(n, 0, -1))
        t_ins = time_algorithm(insertion_sort, worst[:])
        t_mrg = time_algorithm(merge_sort, worst[:])
        print(f"{n:>8} | {t_ins:>14.4f} | {t_mrg:>10.4f} | "
              f"{(t_ins / max(t_mrg, 1e-9)):>9.1f}x")
    print("\nDoubling n should roughly 4x insertion sort and ~2.2x merge sort.")


# ---------------------------------------------------------------------------
# 8. RACE-CONDITION DEMONSTRATION
# ---------------------------------------------------------------------------
# Module notes describe a counter being read, incremented, and written
# without synchronization. Here is the actual failure in Python.
#
# CPython's GIL provides *some* protection, but += on a shared mutable
# attribute is NOT atomic; it decomposes to LOAD / ADD / STORE just like
# the pseudocode in the notes.
class UnsafeCounter:
    def __init__(self):
        self.value = 0

    def increment(self):
        # Split the read-modify-write so a thread switch *will* occur between
        # the read and the write. Mirrors the notes' pseudocode:
        #     read counter ; counter = counter + 1 ; write counter
        current = self.value
        # Yield the GIL/scheduler so another thread can interleave here.
        time.sleep(0)
        self.value = current + 1


class SafeCounter:
    def __init__(self):
        self.value = 0
        self._lock = threading.Lock()

    def increment(self):
        with self._lock:
            current = self.value
            time.sleep(0)
            self.value = current + 1


def race_condition_demo(iterations: int = 10_000,
                        threads: int = 4) -> None:
    for counter_class in (UnsafeCounter, SafeCounter):
        counter = counter_class()

        def worker():
            for _ in range(iterations):
                counter.increment()

        thread_pool = [threading.Thread(target=worker) for _ in range(threads)]
        for t in thread_pool: t.start()
        for t in thread_pool: t.join()

        expected = iterations * threads
        actual = counter.value
        print(f"{counter_class.__name__:>14}: expected {expected:>8}, "
              f"got {actual:>8}, lost {expected - actual:>6}")


# ---------------------------------------------------------------------------
# DEMO HARNESS
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    print("=== Linear vs Binary Search ===")
    data_sorted = list(range(1, 1_000_001))     # 1..1,000,000
    target = 987_654
    idx, cmp_lin = linear_search(data_sorted, target)
    print(f"Linear search comparisons : {cmp_lin:,}  (expected ~{target:,})")
    idx, cmp_bin = binary_search_iterative(data_sorted, target)
    print(f"Binary search comparisons : {cmp_bin:,}  "
          f"(expected ~log2 n = {math.log2(len(data_sorted)):.1f})")

    print("\n=== Insertion Sort: best vs worst case ===")
    sorted_input = list(range(1, 1001))
    reverse_input = list(range(1000, 0, -1))
    _ = insertion_sort(sorted_input[:])
    print(f"Sorted (best)   comparisons : "
          f"{insertion_sort(sorted_input[:]):>7,}   ≈ n - 1 = 999")
    print(f"Reversed (worst) comparisons: "
          f"{insertion_sort(reverse_input[:]):>7,}   ≈ n(n-1)/2 = "
          f"{1000 * 999 // 2:,}")

    print("\n=== Merge Sort ===")
    arr = random.sample(range(10_000), 1_000)
    sorted_arr, cmp_merge = merge_sort(arr)
    print(f"Merge sort comparisons on n=1000: {cmp_merge:,}  "
          f"(n log2 n ≈ {int(1000 * math.log2(1000)):,})")
    assert sorted_arr == sorted(arr)

    print("\n=== Towers of Hanoi ===")
    for n in (1, 3, 5, 10):
        moves = towers_of_hanoi(n)
        print(f"n = {n:>2}: {len(moves):>5,} moves   (2^n - 1 = "
              f"{2 ** n - 1})")

    print("\n=== Recursive vs Iterative Factorial ===")
    print(f"factorial_iterative(10) = {factorial_iterative(10)}")
    print(f"factorial_recursive(10) = {factorial_recursive(10)}")
    print("(Recursive version uses Θ(n) stack — try n = 5000 to see the limit.)")

    print("\n=== Empirical Growth Demo ===")
    growth_demo()

    print("\n=== Race Condition Demo ===")
    race_condition_demo()
